source/postprocess/visualization/heat_flux_map.cc

```diff
@@ -0,0 +1,161 @@
+                                          this->get_fe(),
+                                          quadrature_formula,
+                                          update_gradients      | update_values |
+                                          update_normal_vectors |
+                                          update_q_points       | update_JxW_values);
+
+        typename MaterialModel::Interface<dim>::MaterialModelInputs in(fe_face_values.n_quadrature_points, this->n_compositional_fields());
+        typename MaterialModel::Interface<dim>::MaterialModelOutputs out(fe_face_values.n_quadrature_points, this->n_compositional_fields());
+
+        std::vector<Tensor<1,dim> > temperature_gradients (quadrature_formula.size());
+        std::vector<std::vector<double> > composition_values (this->n_compositional_fields(),std::vector<double> (quadrature_formula.size()));
+
+        // loop over all of the surface cells and evaluate the heatflux
+        typename DoFHandler<dim>::active_cell_iterator
+        cell = this->get_dof_handler().begin_active(),
+        endc = this->get_dof_handler().end();
+
+        unsigned int cell_index = 0;
+        for (; cell!=endc; ++cell,++cell_index)
+          if (cell->is_locally_owned())
+            for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+              if (cell->at_boundary(f))
+                {
+                  fe_face_values.reinit (cell, f);
+
+                  // get the various components of the solution, then
+                  // evaluate the material properties there
+                  fe_face_values[this->introspection().extractors.temperature].get_function_gradients (this->get_solution(),
+                      temperature_gradients);
+
+                  for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                    fe_face_values[this->introspection().extractors.compositional_fields[c]].get_function_values(this->get_solution(),
+                        composition_values[c]);
+
+
+                  for (unsigned int i=0; i<fe_face_values.n_quadrature_points; ++i)
+                    {
+                      for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                        in.composition[i][c] = composition_values[c][i];
+                    }
+
+                  this->get_material_model().evaluate(in, out);
+
+
+                  // Calculate the normal conductive heat flux given by the formula
+                  //   j = - k * n . grad T
+
+                  double normal_flux = 0;
+                  for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
+                    {
+                      const double thermal_conductivity
+                        = out.thermal_conductivities[q];
+
+                      normal_flux += -thermal_conductivity *
+                                     (temperature_gradients[q] * fe_face_values.normal_vector(q)) *
+                                     fe_face_values.JxW(q);
```

Comment:
it might actually be even more wrong if you have a cell that has more than one face on the boundary. think about how you want to deal with that.